Documentation Updates (#3)
coderaiser/supertape: README.md

[MITLicenseIMGURL]: https://img.shields.io/badge/License-MIT-yellow.svg

[![supertape](https://asciinema.org/a/Cgc3rDOfZAeDnJSxzEYpPfBMY.svg)](https://asciinema.org/a/Cgc3rDOfZAeDnJSxzEYpPfBMY)

[**Tape**](https://github.com/substack/tape)-inspired [`TAP`](https://testanything.org/)-compatible simplest high speed test runner with superpowers.

----

📼 **Supertape** is a fast, minimal test runner with the soul of **tape**. It's designed to be as compatible as possible with **tape** while still having some key improvements, such as:

- the ability to work with [ESM Modules](https://nodejs.org/api/esm.html) (take a look at [mock-import](https://github.com/coderaiser/mock-import) for mocking and 🎩[ESCover](https://github.com/coderaiser/escover) for coverage)
- a number of [built-in pretty output formatters](#formatting)
- easy [configuration](#list-of-options)
- the ability to [extend](#extending)
- showing colored diff when using the [`t.equal()`](#tequalactual-any-expected-any-message-string-since-v102) and [`t.deepEqual()`](#tdeepequalactual-any-expected-any-message-string-since-v102) assertion operators
- detailed stack traces for `async` functions

> - not needing to specify the number of assertions with [`t.plan()`](https://github.com/substack/tape#tplann)

Reviewer: `t.plan` is not supported but smart timeouts it’s important information.

Author: Yes, this was included to highlight that `t.plan()` doesn't exist in Supertape

Reviewer: ```suggestion
```

I don't think that we should mention this in improvements :)... I never use `t.plan()`, it can have sense for somebody but it's better to add to doesn't contain, if you think it will be useful.

Author: Moved to "doesn't contain" with https://github.com/tommy-mitchell/supertape/commit/37fee4e4dc51135d5f892d35caf613e27f51e94b

#### `--no-check-assertions-count`

Tells 📼 **Supertape** not to check the number of assertions per test case. By default, only one assertion is allowed.

#### `--no-check-scopes`

Tells 📼 **Supertape** not to check test case messages for proper scoping (i.e. in the form `'scope: message'`). By default, messages are expected to be scoped.

#### **<u>Environment Variables</u>:**

#### `SUPERTAPE_TIMEOUT`

Controls the time (in milliseconds) that a test is allowed to run for before timing out. By default, tests time out after 3 seconds (or `3000` ms).

#### `SUPERTAPE_CHECK_DUPLICATES`

Set to `0` to tell 📼 **Supertape** not to check test case messages for duplicates. By default, messages are expected to be unique.

#### `SUPERTAPE_CHECK_ASSERTIONS_COUNT`

Set to `0` to tell 📼 **Supertape** not to check the number of assertions per test case. By default, only one assertion is allowed.

#### `SUPERTAPE_CHECK_SCOPES`

Set to `0` to tell 📼 **Supertape** not to check test case messages for proper scoping (i.e. in the form `'scope: message'`). By default, messages are expected to be scoped.

#### `CI=1`

> TODO: disable progress bar?

Reviewer: When CI detected (we can skip this option since it set by GitHub actions) it disables progress bar int default formatter

#### `SUPERTAPE_PROGRESS_BAR_STACK`

If using the `progress-bar` formatter, sets a flag to show or hide the error stack when an assertion fails. Defaults to `1` (*show*). Set to `0` to hide.

### `node` and `ts-node`

Instead of using the command line, any test file (`.js` or `.ts`) requiring/importing 📼 **Supertape** can be invoked through `node` or `ts-node`, respectively:

```js
// test.js or test.ts
import test from 'supertape';

test('test: from (ts-)node', (t) => {
    t.pass('hello, world!');
    t.end();
});
```

```bash
$ node test.js
# OR
$ ts-node test.ts
```

Options can be specified either by using [environment variables](#environment-variables) or on a per-file basis with `test.init(options)` or `test(message, fn, options)`.

#### **<u>With Environment Variables</u>:**

For example, setting options for the `progress-bar` formatter:

```bash
$ SUPERTAPE_PROGRESS_BAR_MIN=10 node test.js
```

#### **<u>Initialization</u>:**

📼 **Supertape** also provides a function `test.init(options)` that initializes the test engine with the specified options:

```js
const {init} = require('supertape');

init({
    format: 'json-lines',

> });

Reviewer: I don’t think this going to work this way, formatters is set from CLI.

Author: It's the only way to set `format` when using `node`:

![image](https://user-images.githubusercontent.com/1403701/188556891-efe129cd-2732-4220-9c63-bd05782070da.png)

Reviewer: Also `init` is for internal usage only, is it works for you?

Author: Yes, I didn't realize it was internal usage because it's included on the `test` object, as well as `createStream()`:

![image](https://user-images.githubusercontent.com/1403701/188559468-1b481bad-6d6b-4fac-b2fc-595004381d0f.png)

Reviewer: Well it used from the CLI only... OK, it's a good way to init options for all tests :).

```

Any options specified in an individual test take precedence over initialized options.

#### **<u>Per-Test Options</u>:**

Individual test cases take an optional `options` parameter that overrides any previously-specified options:

```js
const test = require('supertape');

test('test: options', (t) => {
    t.pass('setting options');
    t.end();

> }, {format: 'short'});

Reviewer: You can provide other options like `checkAssertionsCount`

Author: That was just to demonstrate usage. The rest of the options are documented [further down](https://github.com/tommy-mitchell/supertape#list-of-options).

Reviewer: The thing is `format` not supported for each test

Author: I'm currently using it in a wrapper (https://github.com/coderaiser/supertape/issues/2#issuecomment-1055224954) to set the format when using `node` or `ts-node`:

```js
// wrapper.js
import {test as _test} from 'supertape';

export const test = (message, fn) => _test(message, fn, {
    format: 'progress-bar',
});
```

```js
// test.js
import test from './wrapper';

test('output: with progress bar', (t) => {
  t.pass('no tap output');
  t.end();
});
```

```
$ node test.js
TAP version 13

1..1
# tests 1
# pass 1

# ✅ ok
```

If I didn't specify the format, the output would instead be:

```
$ node test.js
TAP version 13
# output: with progress bar
ok 1 no tap output

1..1
# tests 1
# pass 1

# ok
```

```

#### **<u>Using a Wrapper</u>:**

Instead of setting the same options on each test in multiple files (or in a `test.init()` call in each file), a custom wrapper can be used that sets common options:

```js
import {test as _test} from 'supertape';

export const test = (message, fn) => _test(message, fn, {
    checkAssertionsCount: false,

>     format: 'progress-bar',

Reviewer: ```suggestion
```

#### **<u>Using Other `TAP` Reporters</u>:**

If you'd like to use another output format instead of the built-in formatters, such as [**`tap-min`**](https://www.npmjs.com/package/tap-min), use the `format: 'tap'` option and pipe the output to the reporter of your choice:

```bash
$ supertape tests/*.js -f tap | tap-min
```

This can also be done by setting [test options on a per-file basis](#per-test-options) if using `node` or `ts-node`.

### Extending

📼 **Supertape** maintains a minimal number of core assertions, leaving out aliases and things like `t.throws()`. Custom extension operators can either be added through the `test.extend(extensions)` function or through options (such as in [`test()`](#per-test-options) or [`test.init()`](#initialization)).

#### **<u>`test.extend(extensions)`</u>:**

The `extend()` function takes a dictionary of functions that each take an object of built-in operators as input, and return an operator function with any number of arguments. The return value of the function is a `test` object with the custom operators added to it. For example, adding a `transform` operator:

```ts

> import {extend} from 'supertape';

Reviewer: ```suggestion
import {extend, Operator, Test} from 'supertape';
```

Author: Added with https://github.com/tommy-mitchell/supertape/commit/6c43cadb748d6f1f41d32aea5bbe23f512b746ef

const test = extend({
    /** Transforms `a` by +1 and `b` by -1. */
    transform: (t: Operator) => (a: number, b: number, message: string = 'should transform') => {
        const {is, output} = t.equal(a + 1, b - 1);
        return {
            is,
            output,
        };
    },
});

> test('test: transform', (t) => {

Reviewer: ```suggestion
test('test: transform', (t: Test) => {
```

Author: I kept `Test` off of `t` because the type of `t` is actually a union between `Test` and a mapped type of the custom extension operators.

Author: I've edited the `.d.ts` file to make the implementation of `extend()` more clear by adding the types `ExtendedTest` and `CustomTest` in https://github.com/tommy-mitchell/supertape/commit/e6c0a81a2f47eb33245b1149bfc991ed7d9e88dd

Reviewer: Would be great to add tests for this

Whether or not to mark a test case as the only one run by the process. Defaults to `false`.

#### `extensions: CustomOperators` (since: `v3.5.0`)

Custom extension operators to use in test cases. Defaults to `{}`. (See: [Extending](#extending))

#### `quiet: boolean` (since: `v3.8.0`)

Whether or not to not report test results. Defaults to `false`.

#### `format: string` (since: `v3.8.1`)

Which output format to use for the test results. The [built-in `tap` formatters](#formatting) are `tap`, `fail`, `progress-bar`, `json-lines`, and `short`. Defaults to `tap`.

When using the CLI, the default formatter is `progress-bar`. This can only be overriden using the `--format [format]` flag.

#### `run: boolean` (since: `v3.8.0`)

Whether or not to run a test case. Defaults to `true`.

#### `checkDuplicates: boolean` (since: `v5.6.0`)

Whether or not to check test messages for duplicates. 📼 **Supertape** expects each message to be unique. Defaults to `true` (was `false` until `v6.0.0`).

#### `checkAssertionsCount: boolean` (since: `v6.8.0`)

Whether or not to check the number of assertions per test case. 📼 **Supertape** expects each test to have only one assertion. Defaults to `true`.

> #### `checkScopes: boolean` (since: `v6.7.0`)

Reviewer: I don’t thin `since version` has sense.

Author: Up to you, it's easy to remove. I think it's nice from a usability perspective of "what has changed in different updates?"

Reviewer: We have [releases](https://github.com/coderaiser/supertape/releases) for this purpose. The thing is only latest major version of 📼**Supertape** is supported, so I don't think there is a need to point when it was added.

Author: Removed with https://github.com/tommy-mitchell/supertape/commit/80cff0ebe71f607033039ae146479bae0d2d89e4

[BuildStatusURL]: https://travis-ci.org/coderaiser/supertape "Build Status"
[CoverageURL]: https://coveralls.io/github/coderaiser/supertape?branch=master
[CoverageIMGURL]: https://coveralls.io/repos/coderaiser/supertape/badge.svg?branch=master&service=github
[MITLicenseURL]: https://opensource.org/licenses/MIT

> [MITLicenseIMGURL]: https://img.shields.io/badge/License-MIT-yellow.svg

Reviewer: ```suggestion
[MITLicenseIMGURL]: https://img.shields.io/badge/License-MIT-green.svg
```

  ```

📼 **Supertape** doesn't contain:

- assertion aliases, making the available operators far more concise
- `es3 code` and lot's of [ponyfills](https://github.com/sindresorhus/ponyfill#how-are-ponyfills-better-than-polyfills)
- `t.throws()`, `t.doesNotThrow()` - use [**tryCatch**](https://github.com/coderaiser/try-catch) or [**tryToCatch**](https://github.com/coderaiser/try-to-catch) with `t.equal()` instead

For a list of all built-in assertions, see [Operators](#operators).

## Install

```
npm i supertape -D
```

## Example

```js
// test.js
const test = require('supertape');

test('hello: world', (t) => {
    t.pass('hello, world!');
    t.end();
});
```

Output:

```
$ npx supertape test.js
TAP version 13

1..1
# tests 1
# pass 1

# ✅ ok
```

## Usage and Configuration

Tests can be run either through the `supertape` CLI or through `node` (or `ts-node` if using **TypeScript**). [Options](#list-of-options) can be specified via the command line, through environment variables, or in individual test files.

### Command Line

📼 **Supertape** provides a CLI that supports running both `CJS` and `ESM` tests:

```

> $ [npx] supertape "tests/test1.js" [options]

Reviewer: I would prefer to keep cli output

Author: Readded with https://github.com/tommy-mitchell/supertape/commit/2805a22063dab5421937735453c867724891245b

> $ [npx] supertape "tests/test1.js" [options]

Reviewer: ```suggestion
$ [npx] supertape [options] "test/**/*.js"
```

### Single Call to `t.end()`

Each test must call the [`t.end()`](#tend-since-v310) operator to denote that it has finished. Only one call to `t.end()` is allowed, and no assertions are allowed after it.

This check cannot be disabled. In addition, 🐊 **Putout** has a rule 🐊 [`remove-useless-t-end`][PutoutTEnd] for ensuring it is only called once. (See: [Putout](#-putout))

[PutoutTEnd]: https://github.com/coderaiser/putout/blob/master/packages/plugin-tape/README.md#remove-useless-t-end

#### **❌ <u>Example of Incorrect Code</u>:**

```js
test('hello: world', (t) => {
    t.end();
    t.end();
    t.pass('hello, world!');
});
```

#### **✅ <u>Example of Correct Code</u>:**

```js
test('hello: world', (t) => {
    t.pass('hello, world!');
    t.end();
});
```

### Duplicate Test Messages

📼 **Supertape** expects all test messages to be unique. This check can be disabled with:

- the [`--no-check-duplicates`](#--no-check-duplicates) command line flag
- the [`SUPERTAPE_CHECK_DUPLICATES`](#supertape_check_duplicates) environment variable (set to `0`)
- the [`checkDuplicates`](#checkduplicates-boolean-since-v560) test option (set to `true`)

> #### **❌ <u>Example of Incorrect Code</u>:**

Reviewer: why `<u>` and capitalized `Incorrect Code`?

Author: The capitalization was for consistency with other headings. I tried to underline it to differentiate it from `####` sized text that isn't a section heading, but turns out GitHub Flavored Markdown uses `<ins></ins>` <ins>for underlining</ins> instead of `<u></u>`.

Author: https://github.com/tommy-mitchell/supertape/commit/58f35bb0c000be757b39681a69a622813fdfacec uses `<ins>`

Reviewer: I don’t think we need underlining.

Author: Removed in https://github.com/tommy-mitchell/supertape/commit/11201f31810820280015f4c120ca730df481620e

### Other Built-In Operators

To simplify the core of 📼 **Supertape**, other operators are maintained in separate packages. The following is a list of all such packages:

|Package|Version|
|-------|-------|
| [`@supertape/operator-stub`](/packages/operator-stub) | [![npm][OperatorStubNPMBadge]][OperatorStubNPMLink] |

[OperatorStubNPMBadge]: https://img.shields.io/npm/v/@supertape/operator-stub.svg?maxAge=86400
[OperatorStubNPMLink]: https://www.npmjs.com/package/@supertape/operator-stub

## 🐊 **Putout**

📼 **Supertape** goes well with 🐊 [**Putout**](https://github.com/coderaiser/putout), which has a number of rules for linting and generating high quality tests.

### Migrating From **tape**

> 🐊 + 📼 = ❤️

You can easily convert your tests from **tape** to 📼 **Supertape** using 🐊 **Putout's** built-in [`@putout/plugin-tape`][PutoutPluginTape]. For a conversion example, check out [this output][PutoutExample].

[PutoutPluginTape]: https://github.com/coderaiser/putout/tree/master/packages/plugin-tape#readme
[PutoutExample]: https://github.com/coderaiser/cloudcmd/commit/74d56f795d22e98937dce0641ee3c7514a79e9e6

> ### **ES-Lint** Rules
> 
> 🐊 **Putout** also provides some **es-lint** rules for 📼 **Supertape** to help ensure high quality tests:

Reviewer: ```suggestion
### **ESLint** Rules

🐊 **Putout** also provides some **ESLint** rules for 📼 **Supertape** to help ensure high quality tests:
```
